### PR TITLE
Improve gcc detection in llvm.

### PR DESCRIPTION
**Pull request by tomgreen66**
Have hit this issue on a few systems now where location of gcc cannot be trusted to give a consistent location for the install prefix.  By querying the gcc execuable itself it seems it provides a better approach.

Errors I would get during llvm build would be:
```
     80847        /apps/spack/opt/linux-rhel7-skylake_avx512/gcc-8.1.0/llvm-11.
              0.0-s4pz3g4h3dladmlwlrldac5eflo44izh/bin/clang      -rdynamic CMa
              keFiles/cmTC_b709c.dir/testCCompiler.c.o  -o cmTC_b709c
  >> 80848    ld: cannot find crtbegin.o: No such file or directory
  >> 80849        ld: cannot find -lgcc
  >> 80850        ld: cannot find -lgcc_s
  >> 80851        clang-11: error: linker command failed with exit code 1 (use
              -v to see invocation)
  >> 80852        gmake[1]: *** [cmTC_b709c] Error 1
```

By applying this patch it helps.

### CODE REVIEW

**Comment by haampie** [comment]
This is very useful! Can you check the merge conflict?

**Comment by naromero77** [comment]
This PR is also adding `flang `which is not a functioning Fortran compiler at this time, it is only a frontend. I don't think it should be included in this PR and in fact we spoke explicitly about not adding it to Spack until it is ready and can at least compile Fortran77.

**Comment by tomgreen66** [comment]
Apologies for leaving in the flang option.  Used to hate people who do more than the ticket describes.  Have edited the branch to reduce change to what it should only fix.

**Review by naromero77 — APPROVED** [review]

**Comment by naromero77** [comment]
no problem. 

Let's give @trws a day or two to look the change over.

File: var/spack/repos/builtin/packages/llvm/package.py
Comment (trws) [inline var/spack/repos/builtin/packages/llvm/package.py]:
```suggestion

```

**Review by trws — CHANGES_REQUESTED** [review]
Love it, just remove the comment line, and possibly add a fallback or error message for if the `install:` line isn't there? I'm almost tempted to actually check for something in that prefix too, but really this is better than where we were and I'd like to see it in soon.

**Comment by tomgreen66** [comment]
I wasnt sure how to handle the reporting of the logic hence I left the assumption `install:` always exist.  Is it better to leave the existing as the initial value for `gcc_prefix` so if `install:` isnt found it defaults back to the old value?  I couldnt see any existing method to output to say its using the `install:` value - do we want to quietly do the logic or assume `install:` method is the default and report if the old value is used.

**Comment by naromero77** [comment]
I think if `install:` is not found we need to throw a hard error, e.g.
```
raise InstallError('llvm: cannot find GCC installation prefix')
```

**Comment by tldahlgren** [comment]
Closing and re-opening the PR to force re-running checks since `style` is still `Waiting for status to be reported`.

**Review by trws — APPROVED** [review]

**Comment by naromero77** [comment]
There is still one flake8 error:
```
var/spack/repos/builtin/packages/llvm/package.py:507: [W291] trailing whitespace
```

**Comment by tomgreen66** [comment]
Hi, apologies for all the confusion with this change.  Fixed up the remaining issue.  Will make sure I take more care in the future.

**Comment by trws** [comment]
No worries, we all do it.  LGTM.

**Review by alalazo — APPROVED** [review]
Merging, since it has the approval of both `llvm` maintainers.

**Pull request merged**
